addons/test_mail/tests/test_mail_composer.py

```diff
@@ -1852,6 +1859,137 @@ def test_mail_composer_delete(self):
+                    expected_attachment_info = []
+                    if test_template_values['attachment_ids'] and test_template_values['attachment_ids'][0] != Command.clear():
+                        expected_attachment_info.append({'name': 'SomeCommonAttachment.txt', 'type': 'text/plain'})
+                    expected_test_reports = test_template_values['report_template_ids']
+                    if expected_test_reports and expected_test_reports[0] != Command.clear():
+                        expected_attachment_info.append(
+                            {'name': f'{"TestReport2" if expected_test_reports[0] == self.test_report_2.id else "TestReport3"} for {record.name}.html',
+                             'type': 'text/plain'},
+                        )
+                    self.assertMailMailWEmails(
+                        [email_normalize(record.email_from)],
+                        'sent',
+                        author=self.env.user.partner_id,
+                        content=expected_body,
+                        mail_message=record.message_ids[0],
+                        email_values={
+                            'attachments_info': expected_attachment_info,
+                            'body': expected_body,
+                            'email_from': record.user_id.email_formatted,
+                            'subject': expected_subject,
+                        },
+                    )
+
+        # expect duplicates
+        for with_composer_attachment, template_changes in zip([False, True], [[], [same_attachments]] * 2):
+            with self.subTest(with_attachments=with_composer_attachment, template_changes=template_changes):
+                test_template_values = dict(base_template_values)
+                for change in template_changes:
+                    test_template_values.update(change)
+                self.template.write(test_template_values)
+                # reset records to have the same recipient
+                for record_n, record in enumerate([self.test_records[0], self.test_records[1]]):
+                    record.write(base_record_values[record_n])
+                composer = _instanciate_composer()
+                if with_composer_attachment:
+                    composer.attachment_ids = attachment
+                with self.mock_mail_gateway(mail_unlink_sent=False), self.mock_mail_app():
+                    composer._action_send_mail()
+                self.assertEqual(len(self._mails), 1, 'Should have sent 1 email, and skipped a duplicate.\n'
+                                 f'values: {test_template_values}, composer attachment: {with_composer_attachment}')
+                expected_attachment_info = [{
+                    'name': 'SomeCommonAttachment.txt',
+                    'raw': b'ExtraData',
+                    'type': 'text/plain',
+                }]
+                self.assertMailMailWEmails(
+                    [],
```

Comment:
Thing is I don't know in advance which one will be sent.
I now check that one is sent and one is cancelled.

Reply:
Isn't it possible to use `assertMailMailWRecord` then ? it uses a record instead of an email to find the related mail, then performs the same checks on the mail and outgoing email content.

Reply:
Talking about "don't know which record will be mailed", isn't it deterministic ? Did not check code, but either following given res_ids, either doing a search (hence using order) ? If it is not deterministic, could be interesting to find why as we try to hunt not deterministic code ^^

Reply:
I didn't want to add that in the test as it's not true that we want to ensure that the first one is sent instead of the second one.
But it wouldn't be the only test so it's fine, I went with assertMailMailWRecord.

Also the attachments were not properly checked before this apparently so good thing :D